ekncontent/ekncontent/eknc-domain.c

```diff
@@ -739,40 +739,111 @@ static void
 query_state_free (gpointer data)
 {
   QueryState *state = data;
-  g_slist_free_full (state->models, g_object_unref);
+  g_clear_pointer (&state->models, g_ptr_array_unref);
   g_slice_free (QueryState, state);
 }
 
-void
+typedef void (*ScatterTaskCountdownCompleteCallback) (GError *error, gpointer user_data);
+
+typedef struct {
+  guint remaining;
+  ScatterTaskCountdownCompleteCallback callback;
+  gpointer data;
+  GError *error;
+} ScatterTaskCountdown;
+
+static ScatterTaskCountdown *
+scatter_task_countdown_new (guint remaining,
+                            ScatterTaskCountdownCompleteCallback callback,
+                            gpointer data)
+{
+  ScatterTaskCountdown *countdown = g_new0 (ScatterTaskCountdown, 1);
+  countdown->remaining = remaining;
+  countdown->callback = callback;
+  countdown->data = data;
+
+  return countdown;
+}
+
+static void
+scatter_task_countdown_subtask_completed (ScatterTaskCountdown **countdown_ptr,
+                                          GError *error)
+{
+  ScatterTaskCountdown *countdown = *countdown_ptr;
+
+  /* If we received an error we should still continue to countdown, but
+   * store the error so that the caller can deal with it */
+  if (!countdown->error)
+    countdown->error = error;
+  else
+    g_clear_error (&error);
```

Comment:
You won't need to do this here if you follow my suggestion about error ownership